### PR TITLE
balance, httputil: fix the http schema used when sending a `GET` request to server

### LINKED ISSUE #602: deploy cluster with TiProxy by operator in master version, TiProxy check TiDB status port failure
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
as title

### 3. What did you see instead (Required)
send a https request, actually do not set any TLS config
https://github.com/pingcap/tiproxy/blob/8551648442e35041d8684aa839c70cb1535006e5/lib/util/security/cert.go#L41-L54
```
[2024/07/22 03:17:52.937 +00:00] [INFO] [main.cert] [security/cert.go:141] [require certificates to secure clients connections, disable TLS] [tls=server] [cfg="{\"min-tls-version\":\"1.2\"}"]
[2024/07/22 03:17:52.937 +00:00] [INFO] [main.cert] [security/cert.go:141] [require certificates to secure clients connections, disable TLS] [tls=server] [cfg="{\"min-tls-version\":\"1.2\"}"]
[2024/07/22 03:17:52.937 +00:00] [INFO] [main.cert] [security/cert.go:227] [no CA to verify server connections, disable TLS] [tls=client] [cfg="{\"min-tls-version\":\"1.2\"}"]
[2024/07/22 03:17:52.937 +00:00] [INFO] [main.cert] [security/cert.go:227] [no CA to verify server connections, disable TLS] [tls=client] [cfg="{\"min-tls-version\":\"1.2\"}"]
```
from log, check tidb status port using https
```
[2024/07/22 03:17:52.936 +00:00] [INFO] [main.config] [config/config.go:39] [current config]  [cfg="{\"proxy\":{\"addr\":\"0.0.0.0:6000\",\"pd-addrs\":\"cluster-pd:2379\",\"frontend-keepalive\":{\"enabled\":true},\"backendealthy-keepalive\":{\"enabled\":true,\"idle\":60000000000,\"cnt\":5,\"intvl\":3000000000,\"timeout\":15000000000},\"backend-unhealthy-keepalive\":{\"enabled\":true,\"idle\":10000000000,\"cnt\":5,\"intvl\":1000000000,\"timet\":5000000000},\"graceful-close-conn-timeout\":15},\"api\":{\"addr\":\"0.0.0.0:3080\"},\"advance\":{\"ignore-wrong-namespace\":true},\"workdir\":\"/var/lib/tiproxy/work\",\"security\":{\"server-tls\":{\"min-tls-version\":1.2\"},\"server-http-tls\":{\"min-tls-version\":\"1.2\"},\"cluster-tls\":{\"min-tls-version\":\"1.2\"},\"sql-tls\":{\"min-tls-version\":\"1.2\"}},\"log\":{\"encoder\":\"tidb\",\"level\":\"info\",\"log-file\":{\"filename\":/var/lib/tiproxy/log/tiproxy.log\",\"max-size\":300,\"max-days\":3,\"max-backups\":3}},\"balance\":{\"policy\":\"resource\"},\"ha\":{}}"]
.
.
.
[2024/07/22 03:18:37.956 +00:00] [INFO] [main.nsmgr.observer] [observer/backend_observer.go:216] [update backend] [namespace=default] [backend_addr=cluster-tidb-0.cluster-tidb-peer.oltp-utf-ext-stats-tps-7606307-1-191.svc:00] [prev=healthy] [cur="down, err: Get \"https://cluster-tidb-0.cluster-tidb-peer.oltp-utf-ext-stats-tps-7606307-1-191.svc:10080/status\": tls: first record does not look like a TLS handshake: connect status port failed"]
[2024/07/22 03:18:37.956 +00:00] [INFO] [main.nsmgr.observer] [observer/backend_observer.go:216] [update backend] [namespace=default] [backend_addr=cluster-tidb-1.cluster-tidb-peer.oltp-utf-ext-stats-tps-7606307-1-191.svc:00] [prev=healthy] [cur="down, err: Get \"https://cluster-tidb-1.cluster-tidb-peer.oltp-utf-ext-stats-tps-7606307-1-191.svc:10080/status\": tls: first record does not look like a TLS handshake: connect status port failed"]
[2024/07/22 03:18:59.173 +00:00] [ERROR] [main.proxy.conn.be] [backend/backend_conn_mgr.go:291] [get backend failed] [connID=0] [client_addr=10.200.100.4:49252] [addr=0.0.0.0:6000] [ns=default] [duration=10.904556379s] [la_err="No available TiDB instances, please make sure TiDB is available"]
[2024/07/22 03:18:59.173 +00:00] [WARN] [main.proxy.conn] [client/client_conn.go:64] [new connection failed] [connID=0] [client_addr=10.200.100.4:49252] [addr=0.0.0.0:6000] [capability=CLIENT_LONG_PASSWORD|CLIENT_LONG_FLAGLIENT_LOCAL_FILES|CLIENT_PROTOCOL_41|CLIENT_TRANSACTIONS|CLIENT_SECURE_CONNECTION|CLIENT_MULTI_RESULTS|CLIENT_PLUGIN_AUTH] [proxy-protocol=false] [backend_addr=] [quit_source="get backend fail"] [error="No available TiDB itances, please make sure TiDB is available"]
[2024/07/22 03:19:11.705 +00:00] [ERROR] [main.proxy.conn.be] [backend/backend_conn_mgr.go:291] [get backend failed] [connID=1] [client_addr=10.200.100.4:47092] [addr=0.0.0.0:6000] [ns=default] [duration=13.438331753s] [la_err="No available TiDB instances, please make sure TiDB is available"]
```

GET http request is work
```
[root@cluster-tiproxy-0 log]# curl "http://cluster-tidb-0.cluster-tidb-peer.oltp-utf-ext-stats-tps-7606307-1-191.svc:10080/status"
{"connections":0,"version":"8.0.11-TiDB-v8.2.0-alpha","git_hash":"41ed0e508d901753f4fac3bd3aae9b13bd202481"}[root@cluster-tiproxy-0 log]#
```
### 4. What is your version? (Required)

```
[2024/07/22 03:17:52.936 +00:00] [INFO] [main] [server/server.go:207] [Welcome to TiProxy.]  ["Release Version"=v1.1.0-9-g3e160fc] ["Git Commit Hash"=3e160fc8bad0a84ac1a59b2c561daf6d4039008d] ["Git Branch"=HEAD] ["UTC BuilTime"="2024-07-19 05:07:20"] [GoVersion=go1.21.10] [OS=linux] [Arch=amd64]
```

### PR DESCRIPTION
**Pull request by JQWong7**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiproxy/issues/602

Problem Summary:

What is changed and how it works: 
1. before send a http/https request, check TLS config loaded or not. if not nil, url will use https schema.
2. init `http.Client` with `certManager.ClusterTLS()`, which will get the latest TLS config if user online import TLS config.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Notable changes

- [ ] Has configuration change
- [ ] Has HTTP API interfaces change
- [ ] Has tiproxyctl change
- [ ] Other user behavior changes

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/util/httputil/http.go [hunk 1]
```diff
@@ -4,24 +4,49 @@
...
 	pnet "github.com/pingcap/tiproxy/pkg/proxy/net"
 )
 
-func Get(httpCli http.Client, addr, path string, b backoff.BackOff) ([]byte, error) {
+type Client struct {
+	cli          *http.Client
+	getTLSConfig func() *tls.Config
+}
+
+func NewHTTPClient(certManager *cert.CertManager) *Client {
```
Comment (djshow832) [inline pkg/util/httputil/http.go:24]: You can just pass the function to simplify import.

File: pkg/util/httputil/http.go [hunk 2]
```diff
@@ -4,24 +4,49 @@
...
+	return &Client{
+		cli: http.DefaultClient,
+		getTLSConfig: func() *tls.Config {
+			return certManager.ClusterTLS()
+		},
+	}
+}
+
+func (c *Client) SetTimeout(t time.Duration) {
+	c.cli.Timeout = t
```
Comment (djshow832) [inline pkg/util/httputil/http.go:37]: No, the `cli` is used concurrently. You can copy a `cli` and set the timeout.

File: pkg/balance/metricsreader/backend_reader.go [hunk 3]
```diff
@@ -61,14 +60,15 @@ type BackendReader struct {
 	election          elect.Election
 	cfgGetter         config.ConfigGetter
 	backendFetcher    TopologyFetcher
-	httpCli           *http.Client
+	httpCli           *httputil.Client
+	httpSchema        string
```
Comment (djshow832) [inline pkg/balance/metricsreader/backend_reader.go:64]: Revert this

File: pkg/util/httputil/http.go (hunk 2 above)
Comment (River2000i) [reply pkg/util/httputil/http.go:37]: do you mean expose the `Cli` for other package set the timeout?

Comment (djshow832) [reply pkg/util/httputil/http.go:31]: Either is fine.
E.g. you can pass `timeout` to `Get()`, `Get()` copies `cli` and sets `timeout`.

File: pkg/server/server.go [hunk 4]
```diff
@@ -5,7 +5,7 @@ package server
 
 import (
 	"context"
-	"net/http"
+	"github.com/pingcap/tiproxy/pkg/util/httputil"
```
Comment (djshow832) [inline pkg/server/server.go:8]: Refine the import

File: pkg/util/httputil/http.go [hunk 5]
```diff
@@ -4,24 +4,43 @@
...
+}
+
+func Get(httpCli Client, addr, path string, b backoff.BackOff) ([]byte, error) {
 	schema := "http"
-	if v, ok := httpCli.Transport.(*http.Transport); ok && v != nil && v.TLSClientConfig != nil {
+	if tlsConfig := httpCli.getTLSConfig(); tlsConfig != nil {
 		schema = "https"
+		httpCli.cli.Transport = &http.Transport{TLSClientConfig: tlsConfig}
```
Comment (djshow832) [inline pkg/util/httputil/http.go:38]: It's fine to initialize it in `NewHTTPClient`. `tlsConfig` updates itself after reloading config.

File: pkg/util/httputil/http_test.go [hunk 6]
```diff
@@ -28,29 +29,29 @@ func TestHTTPGet(t *testing.T) {
 	wg.Run(func() {
 		_ = statusServer.Serve(statusListener)
 	})
-
-	httpCli := *http.DefaultClient
-	httpCli.Timeout = time.Second
+	certManager := cert.NewCertManager()
+	httpCli := NewHTTPClient(certManager.ClusterTLS)
```
Comment (djshow832) [inline pkg/util/httputil/http_test.go:33]: No need to create `CertManager`, just mock a function.

File: pkg/manager/namespace/manager.go [hunk 7]
```diff
@@ -9,7 +9,7 @@ package namespace
 import (
 	"context"
 	"fmt"
-	"net/http"
+	"github.com/pingcap/tiproxy/pkg/util/httputil"
```
Comment (djshow832) [inline pkg/manager/namespace/manager.go:12]: Refine the import

File: pkg/balance/metricsreader/backend_reader_test.go [hunk 8]
```diff
@@ -180,7 +182,9 @@ func TestReadBackendMetric(t *testing.T) {
 	port := httpHandler.Start()
 	t.Cleanup(httpHandler.Close)
 	addr := net.JoinHostPort("127.0.0.1", strconv.Itoa(port))
-	br := NewBackendReader(lg, nil, http.DefaultClient, nil, cfg)
+	certManager := cert.NewCertManager()
+	cli := httputil.NewHTTPClient(certManager.ClusterTLS)
+	br := NewBackendReader(lg, nil, cli, nil, cfg)
```
Comment (djshow832) [inline pkg/balance/metricsreader/backend_reader_test.go:187]: No need to create `CertManager`, just mock a function that simply returns nil.

File: pkg/util/httputil/http.go [hunk 9]
```diff
@@ -4,24 +4,42 @@
...
+func NewHTTPClient(getTLSConfig func() *tls.Config) *Client {
+	return &Client{
+		cli: &http.Client{
+			Transport: &http.Transport{TLSClientConfig: getTLSConfig()},
+		},
+		getTLSConfig: getTLSConfig,
+	}
+}
+
+func Get(c Client, addr, path string, b backoff.BackOff, timeout time.Duration) ([]byte, error) {
```
Comment (djshow832) [inline pkg/util/httputil/http.go:32]: Why not make it a member of `Client`?
`func (client *Client) Get(...) ...`

File: pkg/util/httputil/http.go [hunk 10]
```diff
@@ -4,24 +4,42 @@
...
+	httpCli := NewHTTPClient(c.getTLSConfig)
```
Comment (djshow832) [inline pkg/util/httputil/http.go:33]: I mean, the server shares the same `httpCli` but copies `cli`. The `cli` is shared and `Timeout` is written concurrently.

File: pkg/balance/metricsreader/backend_reader.go [hunk 11]
```diff
@@ -21,7 +20,7 @@ import (
 	"github.com/pingcap/tiproxy/lib/config"
 	"github.com/pingcap/tiproxy/lib/util/waitgroup"
 	"github.com/pingcap/tiproxy/pkg/manager/elect"
-	"github.com/pingcap/tiproxy/pkg/util/httputil"
+	http "github.com/pingcap/tiproxy/pkg/util/httputil"
```
Comment (djshow832) [inline pkg/balance/metricsreader/backend_reader.go:23]: Let's rename the package `httputil` to `http`.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**